alacritty.yml

```diff
@@ -450,6 +468,27 @@
     # binding section.
     #modifiers: None
 
+# Regex hints
+#
+# Terminal hints can be used to find text in the visible part of the terminal
+# and pipe it to other applications.
+hints:
+  # Keys used for the hint labels
+  #alphabet: "jfkdls;ahgurieowpq"
+
+  # List with all available hints
+  #
+  # The fields `command`, `binding.key` and `binding.mods` accept the same
+  # values as they do in the `key_bindings` section.
+  #
+  # Example:
+  #  - regex: "alacritty/alacritty#\\d*"
+  #    command: firefox
+  #    binding:
+  #      key: G
+  #      mods: Control|Shift
+  #enabled: []
```

Comment:
@nixpulvis Honestly the name of this variable is one of the things I've struggled most with for this feature. Since `hint` is already used for the parent I'm not sure what the best choice here would be.

I suppose `enabled` *works*, but it's not really saying much at all.

Reply:
I think there are larger issues to address first.

I think it's wrong to have new keybindings specified anywhere other than the `key_bindings` section.

Furthermore, I believe this is in fact a new "mode", albeit a very lightweight one. That said, I understand why we don't really want to add modes, because numerous bindings with selections like `mode: ~Vi|~Search` start to get unwieldy, but I think that's an issue we're going to need to address at some point. For now, perhaps we roll this into `Vi` mode?

Reply:
> I think it's wrong to have new keybindings specified anywhere other than the key_bindings section.

I don't think there's any other way to do this wrt how this feature is supposed to work.

I've originally put this in the keybindings section, but that makes it impossible to build on this feature in the future. Which I think would be a massive mistake. So really there's not really any alternative.

> Furthermore, I believe this is in fact a new "mode", albeit a very lightweight one.

That's rather pointless, since no other keybindings will ever be triggered while this is active. So there's no reason to make this a dedicated mode. That just doesn't make any sense at all. Making it a mode just because it feels like a mode rather than how it actually functions isn't really productive.

Reply:
Also...

I'm finding myself trying to get out of this mode by pressing it's activation binding again (like `Vi` mode), however it seems to operate like `Search` mode. I *really* wish these all worked the same way. Personally, after months of using these new features, I still find the way I enter/exit them very confusing.

I would propose that <kbd>ESC</kbd> always exits one mode up, like it currently does from a hint while in a non-vi search. Additionally, I'd make both search and hints exit on repeated activation bindings. I know this isn't how <kbd>ctrl</kbd>-<kbd>f</kbd> works in the browser, but I feel it would be an improvement. Especially since it would allow new searches more easily. But I'm getting a bit off task now.

The point is, we need to make it more consistent how to switch "modes", where I mean mode as in anything that changes what my keystrokes will do next.

On that subject, I also believe we should activate/update the vi-mode overlay with a symbol indicating we're in hint-mode to further help with this issue.

Reply:
> I would propose that ESC always exits one mode up, like it currently does from a hint while in a non-vi search. 

I mean one is free to configure that if they prefer so. But as previously mentioned I much prefer it this way.

> Additionally, I'd make both search and hints exit on repeated activation bindings. I know this isn't how ctrl-f works in the browser, but I feel it would be an improvement.

I honestly don't see how that would make a lot of sense. At least I would never intuitively go to pressing those keys again to get out of it. Especially with the existing behavior in browsers etc.

> On that subject, I also believe we should activate/update the vi-mode overlay with a symbol indicating we're in hint-mode to further help with this issue.

The hints should be enough of an indication? I fail to see how the hints wouldn't be indication enough to show you that you're currently in hint mode.

I've considered stripping all other colors during hint mode to make them easier to identify, but I personally find that switch of the entire viewport a bit disorienting.

Reply:
> That's rather pointless, since no other keybindings will ever be triggered while this is active. So there's no reason to make this a dedicated mode. That just doesn't make any sense at all. Making it a mode just because it feels like a mode rather than how it actually functions isn't really productive.

We ran into this with search too. Maybe we need another word for what it is, because mode is a more powerful concept... but it *is* something that changes the behavior of my next keys, and it's really important to me that entering and exiting this kind of state is consistent across the application.

Reply:
OK, here's a proposal for something I think makes a bit more sense, granted it's a small step up in YAML complexity.

```yaml
match_commands:
  - &URL
    regex: "(.*)://.*"
    command:
      program: open
      args: []
  - &ALACRITTY
    regex: "alacritty/alacritty#\\d*"
    command:
      program: firefox
      args: --safe-mode

hint_alphabet: "jfkdls;ahgurieowpq"
key_bindings:
  - key: G
    mods: Control|Shift
    hint_command: *ALACRITTY
  - key: U
    mods: Control|Shift
    hint_command: *URL
mouse_bindings:
  - mouse: Left
    match_command: *URL
```

I feel it's worth it to keep the `key_bindings` all together. However, it does make the key `match_commands` possibly unused (if not referenced), which is the biggest downside to this approach in my opinion.

I'm not attached to the naming of `match_commands` in general, I'm just personally having trouble with the name "hint" for something that performs an action. But that's a separate smaller naming thing.

Reply:
That doesn't make much sense though. Not only does it force us to rely on yaml features that are way to complicated for anyone's good, which makes this basically unusable, there's also no way for us to identify these.

So basically we're just telling users to copy/paste things and duplicate them everywhere. Which is absolutely garbage if you ask me. That's just a giant mess.

Reply:
I see your point. Perhaps the best thing we can do is to just add a note about the other binding source in the key and mouse binding documentation.

How should I expect overlapping bindings to work with hints? Should they just add to the list of actions that get performed? Or overwrite the whole thing, preventing other actions from taking place?

Reply:
> How should I expect overlapping bindings to work with hints? Should they just add to the list of actions that get performed? Or overwrite the whole thing, preventing other actions from taking place?

Overlapping bindings defined by the user are just duplicated, so it shouldn't matter. Though currently the hint bindings do not override default bindings.